doc/content/design/add-qcow-tool-for-vdi-import-export.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Add qcow tool to allow VDI import/export
+layout: default
+design_doc: true
+revision: 1
+status: proposed
+---
+
+# Introduction
+
+At XCP-ng, we are working on overcoming the 2TiB limitation for VM disks while
+preserving essential features such as snapshots, copy-on-write capabilities, and
+live migration.
+
+To achieve this, we are introducing Qcow2 support in SMAPI and the blktap driver.
+With the alpha release, we can:
+    - Create a VDI
+    - Snapshot it
+    - Export and import it to/from XVA
+    - Perform full backups
+
+However, we currently cannot export a VDI to a Qcow2 file, nor import one.
+
+The purpose of this design proposal is to outline a solution for implementing VDI
+import/export in Qcow2 format.
+
+# Design Proposal
+
+The import and export of VHD-based VDIs currently rely on *vhd-tool*, which is
+responsible for streaming data between a VDI and a file. It supports both Raw and
+VHD formats, but not Qcow2.
+
+There is an existing tool called [qcow-tool](https://opam.ocaml.org/packages/qcow-tool/)
+originally packaged by MirageOS. It is no longer actively maintained, but it can
+produce Qcow files readable by QEMU.
+
+Currently, *qcow-tool* does not support streaming, but we propose to add this
+capability. This means replicating the approach used in *vhd-tool*, where data is
+pushed to a socket.
+
+We have contacted the original developer, David Scott, and there are no objections
+to us maintaining the tool if needed.
+
+Therefore, the most appropriate way to enable Qcow2 import/export in XAPI is to
+add streaming support to `qcow-tool`.
+
+# XenAPI changes
+
+## The workflow
+
+- The export and import of VDIs are handled by the XAPI HTTP server:
+  - `GET /export_raw_vdi`
+  - `PUT /import_raw_vdi`
+- The corresponding handlers are `Export_raw_vdi.handler` and
+  `Import_raw_vdi.handler`.
+- Since the format is checked in the handler, we need to add support for `Qcow2`,
+  as currently only `Raw`, `Tar`, and `Vhd` are supported.
+- This requires adding a new type in the `Importexport.Format` module and a new
+  content type: `"application/qcow2"`.
```

Comment:
There's no official type for qcow2, so I recommend against using a mime type that's not prepended with `x-`. Following [qcow's mime type](https://www.digipres.org/formats/mime-types/#application/x-qemu-disk), I recommend `application/x-qemu-disk-2`, removing '-2' might also be acceptable.

Reply:
I will update that thanks. I'm also trying to better understand how `vhd-tool serve` and `vhd-tool stream` are working because it is not clear yet what is the Unix file descriptor that is used and how it works within the XAPI. So I'm currently having a closer look to the code and I will update the section.
Thanks for the comment about mimetype :+1: